soh/soh/Enhancements/randomizer/fishsanity.cpp

```diff
@@ -321,12 +336,171 @@ namespace Rando {
+            !(actor->flags & ACTOR_FLAG_UPDATE_WHILE_CULLED) && fs->GetOverworldFishShuffled()) {
+            fishGroupCounter = 0;
+        }
+    }
+
+    void Fishsanity::OnSceneInitHandler(int16_t sceneNum) {
+        if (sceneNum == SCENE_ZORAS_DOMAIN) {
+            fishGroupCounter = 0;
+        }
+    }
+
+    void Fishsanity::OnVanillaBehaviorHandler(GIVanillaBehavior id, bool* should, void* refActor) {
+        Actor* actor = static_cast<Actor*>(refActor);
+        auto fs = OTRGlobals::Instance->gRandoContext->GetFishsanity();
+
+        if (id == VB_BOTTLE_ACTOR && actor->id == ACTOR_EN_FISH && fs->GetOverworldFishShuffled()) {
+            FishIdentity fish = OTRGlobals::Instance->gRandomizer->IdentifyFish(gPlayState->sceneNum, actor->params);
+            if (fish.randomizerCheck != RC_UNKNOWN_CHECK && !Flags_GetRandomizerInf(fish.randomizerInf)) {
+                Flags_SetRandomizerInf(fish.randomizerInf);
+                actor->parent = &GET_PLAYER(gPlayState)->actor;
+                *should = false;
+            }
+        }
+    }
+
+    void Fishsanity::RegisterHooks() {
+        static uint32_t onActorInitHook = 0;
+        static uint32_t onFlagSetHook = 0;
+        static uint32_t onActorUpdateHook = 0;
+        static uint32_t onSceneInitHook = 0;
+        static uint32_t onVanillaBehaviorHook = 0;
+
+        GameInteractor::Instance->RegisterGameHook<GameInteractor::OnLoadGame>([](int32_t fileNum) {
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnActorInit>(onActorInitHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnFlagSet>(onFlagSetHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnActorUpdate>(onActorUpdateHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnSceneInit>(onSceneInitHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnVanillaBehavior>(onVanillaBehaviorHook);
+
+            onActorInitHook = 0;
+            onFlagSetHook = 0;
+            onActorUpdateHook = 0;
+            onSceneInitHook = 0;
+
+            if (!IS_RANDO || RAND_GET_OPTION(RSK_FISHSANITY) == RO_FISHSANITY_OFF) return;
+
+            FSi->InitializeFromSave();
+
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnActorInit>(Fishsanity::OnActorInitHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnFlagSet>(Fishsanity::OnFlagSetHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnActorUpdate>(Fishsanity::OnActorUpdateHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneInit>(Fishsanity::OnSceneInitHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnVanillaBehavior>(Fishsanity::OnVanillaBehaviorHandler);
+        });
```

Comment:
I think that it'd be better to integrate this into the main rando `OnLoadGame` hook and not register another `OnLoadGame` hook. I can't come up with a clean way of doing this right now but it'd be great if eventually everything rando related is exclusively in the `OnLoadGame` hook inside `hook_handlers.cpp`.